Remove VAMP plugin support.

rryan opened the pull request:
- Removes VAMP entirely.
- Moves QM-DSP to lib/qm-dsp.
- Introduces AnalyzerPlugin, a class that individual algorithms for extracting features can sub-class -- with implementations for Queen Mary beats/key detectors and SoundTouch.
- AnalyzerBeats / AnalyzerKey use user preferences to decide which AnalyzerPlugin to use.
- Totally removed manual ConfigKey use for beat / key detection preferences -- now using a wrapper class like @daschuer introduced for ReplayGain preferences.

rryan committed:
Delete most VAMP code and move Queen Mary DSP library to lib/qm-dsp.

rryan committed:
Remove VAMP from NSIS.

rryan committed:
Remove VAMP from Travis and README.

rryan committed:
Introduce interface for key and beat detection plugins.

rryan committed:
Use AnalyzerPlugin interface in AnalyzerKey/AnalyzerBeats and preferences.

rryan committed:
Add helper for performing mono downmixing, conversion and overlapped
windowing of buffers.

The Queen Mary detectors expect 50% overlapping windows of samples to be
fed to them. They also use double samples instead of float.

rryan committed:
Add AnalyzerQueenMaryKey.

Uses the qm-dsp library to calculate the key mode of the track. This is
a re-write of the VAMP plugin so the parameters used are the same.

rryan committed:
Add AnalyzerQueenMaryBeats and AnalyzerSoundTouchBeats.

Uses SoundTouch and qm-dsp libraries to calculate BPM and do beat
tracking. Replicates the logic of our VAMP plugins so the parameters to
the algorithms are the same.

rryan committed:
Merge branch 'master' into remove-vamp

rryan committed:
Remove VAMP from .travis.yml.

rryan committed:
Merge branch 'master' into remove-vamp

rryan committed:
Remove manual setting/getting of beat detector preferences.

* Introduce BeatDetectionSettings helper class.
* Remove UserSettingsPointer direct usage from DlgPrefBeats,
  AnalyzerBeats, AnalysisFeature, and Upgrade.
* Delete beat_preferences.h.

rryan committed:
Remove manual setting/getting of key detector preferences.

* Introduce KeyDetectionSettings.
* Use KeyDetectionSettings in DlgPrefKey and AnalyzerKey.
* Delete key_preferences.h.

rryan committed:
Don't hard-code plugin ids.

rryan committed:
Use common default values.

rryan committed:
Oops.

daschuer commented:
GitHub: 
`Sorry, we could not display the entire diff because it was too big.`

daschuer commented on build/depends.py:
```diff
@@ -620,11 +672,13 @@ def sources(self, build):
                    "preferences/dialog/dlgprefsound.cpp",
                    "preferences/dialog/dlgprefsounditem.cpp",
                    "preferences/dialog/dlgprefwaveform.cpp",
+                   'preferences/dialog/dlgprefbeats.cpp',
```
Singe quotes, here and in some other places below.

rryan replied:
good catch -- done

daschuer commented:
Is it possible to rebase the external code changes out? 
Something like a second PR with target rryan:remove-vamp that only contains your own code changes? This would allow us to continue to use GitHub for the review.

rryan committed:
Merge branch 'master' into remove-vamp

rryan committed:
Use consistent quotes.

rryan commented:
> Is it possible to rebase the external code changes out? 
> Something like a second PR with target rryan:remove-vamp that only contains your own code changes? This would allow us to continue to use GitHub for the review.

The history is already a little complicated to do that. What about using:

git diff rryan/remove-vamp mixxxdj/master src

Or a graphical difftool? The changes to src/ are not that complicated.

rryan commented:
> The history is already a little complicated to do that.

Or maybe my git rebase skills are too weak :) I can't think of an easy way. Do you have one?

ywwg commented:
You could squash commits that are just basic removal, and squash the other commits that are changes.  Github now supports easy viewing of individual commits inside one PR

ywwg commented:
(git rebase -i for fun editor-window fine-tuning of the squashing)

rryan commented:
> You could squash commits that are just basic removal, and squash the other commits that are changes. Github now supports easy viewing of individual commits inside one PR

Yea, but I don't see how that's different from the situation now (I have a few bulk removal commits and then some smaller ones). Changing the ordering is tricky because there are merge commits with mass removals in master intermingled so reordering goes to hell immediately with git rebase -i.

rryan commented:
Ugh, I'm not good with the fast clicking today :).

rryan commented:
73bd4a5 is a hairy merge of Uwe's mass moves in master with my mass removals in 2147838.

daschuer commented:
OK, 
git diff rryan/remove-vamp upstream/master src > vamp-removal.patch
dose the job for me, I hope I will find time to comment the patch soon.

daschuer commented:
It is actually: 
```git diff d6d95e00d0b656faba03a2b3b58956d90bd93e4e rryan/remove-vamp src > vamp-removal.patch```

daschuer commented:
```
+    return plugins;
+}
+
 AnalyzerBeats::AnalyzerBeats(UserSettingsPointer pConfig)
-        : m_pConfig(pConfig),
-          m_pVamp(NULL),
+        : m_bpmSettings(pConfig),
           m_bPreferencesReanalyzeOldBpm(false),
           m_bPreferencesFixedTempo(true),
           m_bPreferencesOffsetCorrection(false),
```
Naming: 

 AnalyzerBeats::AnalyzerBeats(BeatDetectionSettings pBeatDetectionSettings)
    : m_pBeatDetectionSettings(pBeatDetectionSettings),

daschuer commented:
Mmm..  I am confused. How is BeatDetectionSettings related to UserSettingsPointer?

daschuer commented:
+// static
+QList<AnalyzerPluginInfo> AnalyzerBeats::availablePlugins() {
+    QList<AnalyzerPluginInfo> plugins;
+    plugins.append(AnalyzerSoundTouchBeats::pluginInfo());
+    plugins.append(AnalyzerQueenMaryBeats::pluginInfo());
+    return plugins;
+}
+

This involves heap allocation every call here and in pluginInfo() as well. It is not a real problem here, but IMHO we should at least cache the list. 
It could be also interesting if it is possible with constexp or something to move the list to the text segment since it is compile time const. 


Will this work? 

constexp QList<AnalyzerPluginInfo*> plugins = {AnalyzerSoundTouchBeats::pPluginInfo(), AnalyzerQueenMaryBeats::pPluginInfo()};

Or should we just use a NULL terminated array?

daschuer commented:
```
+template <typename ConvertTo, typename ConvertFrom>
+class DownmixAndOverlapHelper {
+    typedef std::function<bool(ConvertTo* pBuffer, size_t frames)> BlockReadyCallback;
+  public:
+    DownmixAndOverlapHelper() { }
```

It is not obvious what this Helper does. Could you add a comment? 
This does not need to be a template, since the types are always the same.

daschuer commented:
+                // If the block size equals the step size then we don't have to
+                // do any copying.
+                if (m_stepSize == m_blockSize) {
+                    m_bufferWritePosition = 0;
+                } else {

This case is redundant.

daschuer commented:
+    size_t m_blockSize = 0;
+    size_t m_stepSize = 0;

please add a comment that explains the use of these members. 
It looks like you have dropped some useful comments from the original code.

daschuer commented:
m_blockSize and m_stepSize, are responsible to tweak the samples that a re passed to the FFT code, right? Is there a state less fft used, that we have to pass more samples in a block than in one step? 
This seams to be quite ineffective and does not deal well with the "infinity" samples before the block. 

Can we rename the variables to their meaning in the FFT context? Does it make sense to use terms like annalysed sampes, post / previous samples?

daschuer commented:
```
+          m_minBpm(m_bpmSettings.getBpmRangeStartDefault()),
+          m_maxBpm(m_bpmSettings.getBpmRangeEndDefault()),
+          m_banalyzerEnabled(m_bpmSettings.getBpmDetectionEnabledDefault()),
+          m_bfixedtempoEnabled(m_bpmSettings.getFixedTempoAssumptionDefault()),
+          m_boffsetEnabled(m_bpmSettings.getFixedTempoOffsetCorrectionDefault()),
+          m_FastAnalysisEnabled(m_bpmSettings.getFastAnalysisDefault()),
+          m_bReanalyze(m_bpmSettings.getReanalyzeWhenSettingsChangeDefault()) {
     setupUi(this);
 
-    populate();
+    m_availablePlugins = AnalyzerBeats::availablePlugins();
+    for (const AnalyzerPluginInfo& info : m_availablePlugins) {
+        plugincombo->addItem(info.name, info.id);
+    }
+
     loadSettings();
```
loadSettings() call and the initializer list are redundant. we should do only one.

daschuer commented:
+    // WARNING(rryan): Do not hard-code defaults here! Put them in
+    // KeyDetectionSettings.

A warning warns for something ;-). I think you want a NOTE; or just the plain comment.

daschuer commented:
```
+    // TODO(rryan): Above range enabled is not exposed?
+    m_minBpm = m_bpmSettings.getBpmRangeStartDefault();
```
I think we can remove the range enabled feature, since it still uses 0 and 999 anyway. 
I also have not understood why we need the range at all.

daschuer commented:
```
+    inline void set##name##Default() { set##name(get##name##Default()); }   \
```
This seams to be unused. If it is used, I would prefer reset##name##ToDefault() or just ##name##ToDefault()

rryan committed:
Merge branch 'master' into remove-vamp

rryan commented:
> This involves heap allocation every call here and in pluginInfo() as well. It is not a real problem here, but IMHO we should at least cache the list. 
> It could be also interesting if it is possible with constexp or something to move the list to the text segment since it is compile time const.
> 
> Will this work?
> 
> constexp QList plugins = {AnalyzerSoundTouchBeats::pPluginInfo(), AnalyzerQueenMaryBeats::pPluginInfo()};
> 
> Or should we just use a NULL terminated array?

I could improve this -- but does it matter? I don't think so since this code is only called once per from DlgPrefKey/DlgPrefBeats.

rryan committed:
Rename default accessor to setNameToDefault.

rryan committed:
Note, don't warn.

rryan committed:
Drop some comments and remove template in DownmixAndOverlapHelper.

rryan commented:
> This seams to be unused. If it is used, I would prefer reset##name##ToDefault() or just ##name##ToDefault()

SGTM -- I went with "setNameToDefault" to match the "setName" method.

rryan commented:
> I think we can remove the range enabled feature, since it still uses 0 and 999 anyway.

:+1: though I'd rather do this in a separate PR since the goal here is to be a simple refactor.

rryan commented:
> A warning warns for something ;-). I think you want a NOTE; or just the plain comment.

Switched to NOTE.

rryan commented:
> loadSettings() call and the initializer list are redundant. we should do only one.

Hm, I think of it differently. By initializing to the defaults, we ensure that the class is always in a valid state -- rather than pointing at uninitialized memory. Some logic could be inserted between the call to loadSettings that assumes the members are in a valid state.

rryan commented:
> m_blockSize and m_stepSize, are responsible to tweak the samples that a re passed to the FFT code, right? Is there a state less fft used, that we have to pass more samples in a block than in one step? 

Yep that's right -- this is part of the "enframing" process of a short-time fourier transform where you slice the waveform into overlapping windows and take the FFT of each window. 

> Can we rename the variables to their meaning in the FFT context? Does it make sense to use terms like annalysed sampes, post / previous samples?

In my experience, when doing an STFT the terms are usually block or window size and step size / hop size or overlap size (which is the inverse of step size). 

I don't think I've ever heard of analyzed samples and post/previous samples in the FFT context? What do those mean?

rryan commented:
> // If the block size equals the step size then we don't have to
> // do any copying.
> if (m_stepSize == m_blockSize) {
> m_bufferWritePosition = 0;
> } else {
> This case is redundant.

Good point -- removed.

rryan commented:
> please add a comment that explains the use of these members. 
> It looks like you have dropped some useful comments from the original code.

Done

rryan commented:
> It is not obvious what this Helper does. Could you add a comment? 
> This does not need to be a template, since the types are always the same.

Done and removed the template.

rryan commented:
Thanks for reviewing the diff manually!

daschuer commented:
>> This involves heap allocation every call here and in pluginInfo() as well. It is not a real problem here, but IMHO we should at least cache the list. 
>> It could be also interesting if it is possible with constexp or something to move the list to the text segment since it is compile time const.
>> 
>> Will this work?
>> 
>> constexp QList plugins = {AnalyzerSoundTouchBeats::pPluginInfo(), AnalyzerQueenMaryBeats::pPluginInfo()};
>> 
>> Or should we just use a NULL terminated array?

> I could improve this -- but does it matter? I don't think so since this code is only called once per from DlgPrefKey/DlgPrefBeats.

It is not a real problem, just a small CPU saver, so you can leave it like it is. It was just too eye catching and fun to look if C++11 offers a smart solution.

daschuer commented:
>> Can we rename the variables to their meaning in the FFT context? Does it make sense to use terms like annalysed sampes, post / previous samples?

>In my experience, when doing an STFT the terms are usually block or window size and step size / hop size or overlap size (which is the inverse of step size). 

>I don't think I've ever heard of analyzed samples and post/previous samples in the FFT context? What do those mean?

Ah, now I see. So the analysis is a stateless STFT. I did not got that during my first review. 
The new comments are helping, thank you.  

The term blockSize seams to be in some code, but it is still confusing for me. Maybe because I know only the term "Window size" from German STFT explanations. 

And vamp uses is a third name for it, which brings the confusion to the max ;-)  Frame is used differently in Mixxx. 
```config.frameLength = blockSize;```

How about using m_windowSize in Mixxx?

Be-ing commented:
This pull request has not been updated in years so I am closing it to reduce the clutter of open PRs. If this is still relevant and you solve the merge conflicts and would like to have it merged in the future, please reopen it.

daschuer commented:
@uklotzde @rryan 
Is this still relevant? I think yes.

rryan commented:
Yep, someday I may pick this up again. I deeply regret ever allowing VAMP to merge into Mixxx in the first place, so I feel this is my mess to clean up. :)

rryan commented:
I don't need the pull request to be open to remember it though. You can leave it closed.

rryan committed:
Merge branch 'master' into remove-vamp

rryan committed:
Fix the build.

rryan commented:
Merged with master (ouch).

rryan committed:
Reword block to window throughout.

rryan committed:
Don't build modules we don't need from qm-dsp.

rryan commented:
> How about using m_windowSize in Mixxx?

Works for me -- replaced block with window throughout.

rryan committed:
Remove vamppluginadapter.h/cpp.

rryan committed:
Merge remote-tracking branch 'mixxxdj/master' into remove-vamp

rryan committed:
Remove plugins directory from Wix script.

rryan committed:
Remove const qualifiers on AnalyzerPlugin members to avoid deleting the implicit copy assignment constructor.

rryan committed:
Merge remote-tracking branch 'mixxxdj/master' into remove-vamp

rryan committed:
Don't sign non-existent vamp plugins.

beenisss commented:
I've spent a bit of time testing this on macOS 10.13.6. In the current release (and the current beta) there's an obvious difference between the time taken by the two beat analyzers, but for whatever reason if I build the current master or this PR, there isn't. In the current master I still get different analysis results depending on which beat detection I've selected though, whereas with this PR I get the same results regardless, which suggests it's ignoring my choice of beat analyzer (seems to be using QM every time).

Likewise for Enable Fast Analysis, you can usually see that it only scans about the first 1/4 of the track. but this PR doesn't seem to use fast mode regardless of what my settings are.

I'm not actually sure what difference 'Assume constant tempo' and 'Enable Offset Correction' are meant to make, it's never made any apparent difference when I've experimented with them before, so I can't comment on whether those are working or not.

As for actual detection, the returned values are no less reliable than before (though from the above it appears I'm only testing one beat analyzer.)

rryan committed:
Merge remote-tracking branch 'mixxxdj/master' into remove-vamp

rryan commented:
Thank you for testing @benis 🙏. I'll try to reproduce your findings and fix it.

rryan committed:
Disable key analysis options when key analysis is disabled.

rryan committed:
Disable fast analysis when non-constant beatgrids are enabled.

We have to analyze the entire song for beats if we are not assuming the detected
BPM applies to the whole track.

rryan committed:
Re-implement fast analysis for AnalyzerBeats and AnalyzerKey.

Uses a 1-minute cutoff to match the previous implementation.

rryan commented:
Indeed fast analysis wasn't working -- re-added that.

I'm finding that the detected BPM does change when switching between SoundTouch and Queen Mary -- could you try again and post the mixxx.log? I've added some extra debug output that should help see what's going on.

uklotzde commented:
I'm curious to see if this will solve the multi-threading issues we experienced when using the VAMP plugins in #1624. Only then we are able to unleash the full potential of multi-threaded analysis, which is currently thwarted by synchronization guards.

rryan commented:
@uklotzde yes! If the issue is with VAMP itself then this should help. If it's shared static state within the plugin code itself then we'll have to fix that I guess.

uklotzde commented:
I'm waiting until either #1624 or this PR has been merged to master. I don't want to resolve merge conflicts twice. It's ok to merge this PR first, then I will resolve the conflicts in #1624 before merging.

Be-ing commented:
I'm testing #1624 now. I was wondering why each analysis thread maxes out at ~25% CPU usage for me. Lifting that limit would be great.

Be-ing commented:
After doing a batch analysis, this hangs on shutdown trying to delete the track loaded in a deck from GlobalTrackCache:
```
Debug [Main]: in ~EngineMaster()
Debug [Main]: 740 ms deleting DlgPreferences
Debug [Main]: 835 ms deleting EffectsManager
Debug [Main]: GlobalTrackCache - Evicting track ["/home/be/music/yes/Yussef Kamaal - Black Focus/Yussef Kamaal - Black Focus - 07 Lowrider.flac" | "/home/be/music/yes/Yussef Kamaal - Black Focus/Yussef Kamaal - Black Focus - 07 Lowrider.flac" | 2604] Track(0x2e9cc370)
Debug [Main]: GlobalTrackCache - Skip to save an already evicted track Track(0x2e9cc370)
Debug [Main]: GlobalTrackCache - Deleting Track(0x2e9cc370)
Debug [Main]: GlobalTrackCache - Evicting track ["/home/be/music/yes/Yussef Kamaal - Black Focus/Yussef Kamaal - Black Focus - 09 WingTai Drums.flac" | "/home/be/music/yes/Yussef Kamaal - Black Focus/Yussef Kamaal - Black Focus - 09 WingTai Drums.flac" | 2601] Track(0x2e341980)
Debug [Main]: GlobalTrackCache - Skip to save an already evicted track Track(0x2e341980)
Debug [Main]: GlobalTrackCache - Deleting Track(0x2e341980)
Debug [Main]: 930 ms deleting SettingsManager
Debug [Main]: TrackAnalysisScheduler - Destroying
Debug [Main]: TrackAnalysisScheduler - Destroying
Debug [Main]: Mixxx shutdown complete with code 0
GlobalTrackCache - Deleting Track(0x2e724920)
```

Be-ing commented:
I can reproduce the deadlock reliably by quitting Mixxx while batch analysis is running, even without loading any tracks to any decks.

Be-ing commented:
The deadlock only occurs with batch analysis running on shutdown. If I load an unanalyzed track to a deck then immediately quit Mixxx, there is no deadlock. If I stop batch analysis before quitting Mixxx, there is no deadlock.

uklotzde commented:
The deadlock also happens with #1624, so this issue seems to be unrelated. Investigating...

benis commented:
Seems to work now, did you still need the log file? I've attached it anyway. For the logged session I did:

Analyze with QM
Analyze with SoundTouch
Analyze with QM, fast analysis enabled
Analyze with SoundTouch, fast analysis enabled

QM took less time than SoundTouch in every case, I don't know if that seems right. As before I'm not sure what difference 'Assume constant tempo' and 'Enable Offset Correction' are meant to make, so I haven't tried to test those.

[mixxx.log](https://github.com/mixxxdj/mixxx/files/2713173/mixxx.log)

Be-ing commented:
The preference defaults to using SoundTouch for analysis with an empty settings directory. Why do we even support multiple analyzers?

rryan committed:
Fix default the beat detector to be the QMUL beat tracker.

rryan commented:
> The preference defaults to using SoundTouch for analysis with an empty settings directory. 

Doh, fixed in 260071e1513733b18fddc7d8c0efea25a27a8ac8. 

> Why do we even support multiple analyzers?

It's nice to have the option for multiple so that we can give users more choice (SoundTouch is nicer for slower machines) and experiment with new beat detectors. I'd like to experiment with [Aubio](https://aubio.org/), for example.

Be-ing commented:
Okay, everything seems to work as it does in master. I'm not familiar with this area of the code at all, so I'll leave it to @uklotzde and/or @daschuer to judge that.

Be-ing commented:
> The deadlock also happens with #1624, so this issue seems to be unrelated. Investigating...

Sorry, I accidentally commented on the wrong PR before. I deleted the misplaced comments. The deadlock happened when I was testing #1624.

daschuer reviewed (commented):
Some minor comments.

rryan committed:
Merge remote-tracking branch 'mixxxdj/master' into remove-vamp

rryan commented on src/analyzer/plugins/analyzerqueenmarykey.cpp:
```diff
@@ -0,0 +1,78 @@
+            const auto key = static_cast<ChromaticKey>(iKey);
+            if (key != m_prevKey) {
+                // TODO(rryan) reserve?
+                m_resultKeys.push_back(qMakePair(
```
I agree, though that's beyond the scope of this PR since this is just making a KeyChangeList, which the rest of Mixxx uses.

rryan commented on src/analyzer/plugins/buffering_utils.h:
```diff
@@ -0,0 +1,80 @@
+  private:
+    std::vector<double> m_buffer;
+    // The window size in frames.
+    size_t m_windowSize = 0;
```
I thought we had agreed that we could use initializers if the constructor was defaulted?

daschuer replied:
Oh sorry. I did not notice that we have no constructor in this case.

rryan committed:
Move analyzer constants out of analyzerthread.cpp to analyzer/constants.h.

rryan committed:
Add a constant for the number of seconds to analyze in fast analysis mode.

rryan committed:
Move fast analysis samples calculation to init() and use analyzer/constants.h.

rryan committed:
Use constants and do not use non-static member initializers.

rryan commented on src/analyzer/plugins/analyzersoundtouchbeats.cpp:
```diff
@@ -0,0 +1,48 @@
+        return false;
+    }
+    DEBUG_ASSERT(iLen == 4096 * 2); // TODO(rryan)
+    DEBUG_ASSERT(iLen % 2 == 0);
```
done

rryan commented on src/analyzer/plugins/analyzerqueenmarykey.h:
```diff
@@ -0,0 +1,51 @@
+  private:
+    std::unique_ptr<GetKeyMode> m_pKeyMode;
+    DownmixAndOverlapHelper m_helper;
+    size_t m_currentFrame = 0;
```
done

rryan commented on src/analyzer/plugins/analyzerqueenmarykey.cpp:
```diff
@@ -0,0 +1,78 @@
+        return false;
+    }
+
+    const size_t numInputFrames = iLen / 2;
```
done

rryan commented on src/analyzer/plugins/analyzerqueenmarykey.cpp:
```diff
@@ -0,0 +1,78 @@
+}
+
+bool AnalyzerQueenMaryKey::process(const CSAMPLE* pIn, const int iLen) {
+    DEBUG_ASSERT(iLen % 2 == 0);
```
done

daschuer commented:
Nothing else from the review: LGTM.
I have not tested it though.

rryan commented:
Address sanitizer:
```
=================================================================
==5508==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6220001365e0 at pc 0x000110c77533 bp 0x70000c682c00 sp 0x70000c682bf8
READ of size 4 at 0x6220001365e0 thread T89
==5508==AddressSanitizer: while reporting a bug found another one. Ignoring.
    #0 0x110c77532 in soundtouch::PeakFinder::getPeakCenter(float const*, int) const PeakFinder.cpp:147
    #1 0x110c77a92 in soundtouch::PeakFinder::detectPeak(float const*, int, int) PeakFinder.cpp:236
    #2 0x110c6f224 in soundtouch::BPMDetect::getBpm() BPMDetect.cpp:536
    #3 0x10f8dfecc in mixxx::AnalyzerSoundTouchBeats::finalize() analyzersoundtouchbeats.cpp:45
    #4 0x10f8ab578 in AnalyzerBeats::finalize(std::__1::shared_ptr<Track>) analyzerbeats.cpp:201
    #5 0x10f8bd638 in AnalyzerThread::doRun() analyzerthread.cpp:168
    #6 0x110a30c21 in WorkerThread::run() workerthread.cpp:63
    #7 0x112250aad  (QtCore:x86_64+0x2daad)
    #8 0x7fff6d4d8660 in _pthread_body (libsystem_pthread.dylib:x86_64+0x3660)
    #9 0x7fff6d4d850c in _pthread_start (libsystem_pthread.dylib:x86_64+0x350c)
    #10 0x7fff6d4d7bf8 in thread_start (libsystem_pthread.dylib:x86_64+0x2bf8)

0x6220001365e0 is located 0 bytes to the right of 5344-byte region [0x622000135100,0x6220001365e0)
allocated by thread T89 here:
    #0 0x113c1d22b in wrap__Znam (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x6422b)
    #1 0x110c6ed17 in soundtouch::BPMDetect::getBpm() BPMDetect.cpp:531
    #2 0x10f8dfecc in mixxx::AnalyzerSoundTouchBeats::finalize() analyzersoundtouchbeats.cpp:45
    #3 0x10f8ab578 in AnalyzerBeats::finalize(std::__1::shared_ptr<Track>) analyzerbeats.cpp:201
    #4 0x10f8bd638 in AnalyzerThread::doRun() analyzerthread.cpp:168
    #5 0x110a30c21 in WorkerThread::run() workerthread.cpp:63
    #6 0x112250aad  (QtCore:x86_64+0x2daad)
    #7 0x7fff6d4d8660 in _pthread_body (libsystem_pthread.dylib:x86_64+0x3660)
    #8 0x7fff6d4d850c in _pthread_start (libsystem_pthread.dylib:x86_64+0x350c)
    #9 0x7fff6d4d7bf8 in thread_start (libsystem_pthread.dylib:x86_64+0x2bf8)

Thread T89 created by T0 here:
    #0 0x113c09676 in wrap_pthread_create (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x50676)
    #1 0x1122512c3 in QThread::start(QThread::Priority) (QtCore:x86_64+0x2e2c3)
    #2 0x10f8e119d in TrackAnalysisScheduler::TrackAnalysisScheduler(Library*, int, QSharedPointer<ConfigObject<ConfigValue> > const&, AnalyzerModeFlags) trackanalysisscheduler.cpp:83
    #3 0x10f8e016f in TrackAnalysisScheduler::createInstance(Library*, int, QSharedPointer<ConfigObject<ConfigValue> > const&, AnalyzerModeFlags) trackanalysisscheduler.cpp:58
    #4 0x10ff4ccdf in AnalysisFeature::analyzeTracks(QList<TrackId>) analysisfeature.cpp:133
    #5 0x1102d622e in AnalysisFeature::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) moc_analysisfeature.cc:109
    #6 0x112430d4a in QMetaObject::activate(QObject*, int, int, void**) (QtCore:x86_64+0x20dd4a)
    #7 0x1102e1140 in DlgAnalysis::analyzeTracks(QList<TrackId>) moc_dlganalysis.cc:275
    #8 0x1101f54c9 in DlgAnalysis::analyze() dlganalysis.cpp:141
    #9 0x1102dfd63 in DlgAnalysis::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) moc_dlganalysis.cc:143
    #10 0x112430d4a in QMetaObject::activate(QObject*, int, int, void**) (QtCore:x86_64+0x20dd4a)
    #11 0x1132e549e  (QtWidgets:x86_64+0xfa49e)
    #12 0x1132e5336  (QtWidgets:x86_64+0xfa336)
    #13 0x1132e659e in QAbstractButton::mouseReleaseEvent(QMouseEvent*) (QtWidgets:x86_64+0xfb59e)
    #14 0x11323668e in QWidget::event(QEvent*) (QtWidgets:x86_64+0x4b68e)
    #15 0x1131fb6ac in QApplicationPrivate::notify_helper(QObject*, QEvent*) (QtWidgets:x86_64+0x106ac)
    #16 0x1131fe4ec in QApplication::notify(QObject*, QEvent*) (QtWidgets:x86_64+0x134ec)
    #17 0x1124001b7 in QCoreApplication::notifyInternal2(QObject*, QEvent*) (QtCore:x86_64+0x1dd1b7)
    #18 0x1131fc00f in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool) (QtWidgets:x86_64+0x1100f)
    #19 0x1132565a6  (QtWidgets:x86_64+0x6b5a6)
    #20 0x1132552a4  (QtWidgets:x86_64+0x6a2a4)
    #21 0x1131fb6ac in QApplicationPrivate::notify_helper(QObject*, QEvent*) (QtWidgets:x86_64+0x106ac)
    #22 0x1131fcafb in QApplication::notify(QObject*, QEvent*) (QtWidgets:x86_64+0x11afb)
    #23 0x1124001b7 in QCoreApplication::notifyInternal2(QObject*, QEvent*) (QtCore:x86_64+0x1dd1b7)
    #24 0x1127f6cc9 in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) (QtGui:x86_64+0x2ecc9)
    #25 0x1127deaca in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) (QtGui:x86_64+0x16aca)
    #26 0x119db375d  (libqcocoa.dylib:x86_64+0x2b75d)
    #27 0x119db4020  (libqcocoa.dylib:x86_64+0x2c020)
    #28 0x7fff452bb8f0 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ (CoreFoundation:x86_64h+0xa38f0)
    #29 0x7fff4537530b in __CFRunLoopDoSource0 (CoreFoundation:x86_64h+0x15d30b)
    #30 0x7fff4529e34f in __CFRunLoopDoSources0 (CoreFoundation:x86_64h+0x8634f)
    #31 0x7fff4529d7cc in __CFRunLoopRun (CoreFoundation:x86_64h+0x857cc)
    #32 0x7fff4529d032 in CFRunLoopRunSpecific (CoreFoundation:x86_64h+0x85032)
    #33 0x7fff44587d95 in RunCurrentEventLoopInMode (HIToolbox:x86_64+0x2fd95)
    #34 0x7fff44587a0e in ReceiveNextEventCommon (HIToolbox:x86_64+0x2fa0e)
    #35 0x7fff44587883 in _BlockUntilNextEventMatchingListInModeWithFilter (HIToolbox:x86_64+0x2f883)
    #36 0x7fff42837a72 in _DPSNextEvent (AppKit:x86_64+0x41a72)
    #37 0x7fff42fcde33 in -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] (AppKit:x86_64+0x7d7e33)
    #38 0x7fff4282c884 in -[NSApplication run] (AppKit:x86_64+0x36884)
    #39 0x119db2e0c  (libqcocoa.dylib:x86_64+0x2ae0c)
    #40 0x1123fbd3d in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) (QtCore:x86_64+0x1d8d3d)
    #41 0x1124008b0 in QCoreApplication::exec() (QtCore:x86_64+0x1dd8b0)
    #42 0x10f8350cc in main main.cpp:53
    #43 0x7fff6d1c0014 in start (libdyld.dylib:x86_64+0x1014)

SUMMARY: AddressSanitizer: heap-buffer-overflow PeakFinder.cpp:147 in soundtouch::PeakFinder::getPeakCenter(float const*, int) const
Shadow bytes around the buggy address:
  0x1c4400026c60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c4400026c70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c4400026c80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c4400026c90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c4400026ca0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x1c4400026cb0: 00 00 00 00 00 00 00 00 00 00 00 00[fa]fa fa fa
  0x1c4400026cc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c4400026cd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c4400026ce0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c4400026cf0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c4400026d00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==5508==ABORTING
```

rryan committed:
Use size_t for kWindowSize to be consistent with kStepSize.

Be-ing commented:
Build warning:
```
scons: warning: Ignoring missing SConscript 'lin64_build/vamp-plugins/SConscript'
File "/home/be/sw/mixxx/SConstruct", line 100, in <module>
```
I tried `rm -rf lin64_build .sconf_temp .sconsign*` and the error still appears.

rryan commented:
Hm, happens in master after #1624. Batch analysis with SoundTouch. /cc: @uklotzde 

```
=================================================================
==10350==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x62200000bde0 at pc 0x0001559f73e3 bp 0x7000058ed7a0 sp 0x7000058ed798
READ of size 4 at 0x62200000bde0 thread T63
Warning [Engine]: underflowHappened code: 24
Warning [Engine]: underflowHappened code: 6
    #0 0x1559f73e2 in soundtouch::PeakFinder::getPeakCenter(float const*, int) const PeakFinder.cpp:147
    #1 0x1559f7942 in soundtouch::PeakFinder::detectPeak(float const*, int, int) PeakFinder.cpp:236
    #2 0x1559f42b4 in soundtouch::BPMDetect::getBpm() BPMDetect.cpp:536
    #3 0x1559daf17 in MixxxBpmDetection::getRemainingFeatures() MixxxBpmDetection.cpp:240
    #4 0x1458434ee in _VampPlugin::Vamp::PluginAdapterBase::Impl::getRemainingFeatures(_VampPlugin::Vamp::Plugin*) (libvamp-sdk.so:x86_64+0x34ee)
    #5 0x106da4fad in _VampHost::Vamp::PluginHostAdapter::getRemainingFeatures() (libvamp-hostsdk.so:x86_64+0x3fad)
    #6 0x106db645f in _VampHost::Vamp::HostExt::PluginWrapper::getRemainingFeatures() (libvamp-hostsdk.so:x86_64+0x1545f)
    #7 0x106db645f in _VampHost::Vamp::HostExt::PluginWrapper::getRemainingFeatures() (libvamp-hostsdk.so:x86_64+0x1545f)
    #8 0x102ae212d in mixxx::VampPluginAdapter::getRemainingFeatures() vamppluginadapter.cpp:266
    #9 0x102ad7bac in VampAnalyzer::End() vampanalyzer.cpp:220
    #10 0x102a98681 in AnalyzerBeats::finalize(std::__1::shared_ptr<Track>) analyzerbeats.cpp:199
    #11 0x102aad508 in AnalyzerThread::doRun() analyzerthread.cpp:178
    #12 0x103c36421 in WorkerThread::run() workerthread.cpp:63
    #13 0x10543faad  (QtCore:x86_64+0x2daad)
    #14 0x7fff6d4d8660 in _pthread_body (libsystem_pthread.dylib:x86_64+0x3660)
    #15 0x7fff6d4d850c in _pthread_start (libsystem_pthread.dylib:x86_64+0x350c)
    #16 0x7fff6d4d7bf8 in thread_start (libsystem_pthread.dylib:x86_64+0x2bf8)

0x62200000bde0 is located 0 bytes to the right of 5344-byte region [0x62200000a900,0x62200000bde0)
allocated by thread T63 here:
    #0 0x106e3a22b in wrap__Znam (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x6422b)
    #1 0x1559f3da7 in soundtouch::BPMDetect::getBpm() BPMDetect.cpp:531
    #2 0x1559daf17 in MixxxBpmDetection::getRemainingFeatures() MixxxBpmDetection.cpp:240
    #3 0x1458434ee in _VampPlugin::Vamp::PluginAdapterBase::Impl::getRemainingFeatures(_VampPlugin::Vamp::Plugin*) (libvamp-sdk.so:x86_64+0x34ee)
    #4 0x106da4fad in _VampHost::Vamp::PluginHostAdapter::getRemainingFeatures() (libvamp-hostsdk.so:x86_64+0x3fad)
    #5 0x106db645f in _VampHost::Vamp::HostExt::PluginWrapper::getRemainingFeatures() (libvamp-hostsdk.so:x86_64+0x1545f)
    #6 0x106db645f in _VampHost::Vamp::HostExt::PluginWrapper::getRemainingFeatures() (libvamp-hostsdk.so:x86_64+0x1545f)
    #7 0x102ae212d in mixxx::VampPluginAdapter::getRemainingFeatures() vamppluginadapter.cpp:266
    #8 0x102ad7bac in VampAnalyzer::End() vampanalyzer.cpp:220
    #9 0x102a98681 in AnalyzerBeats::finalize(std::__1::shared_ptr<Track>) analyzerbeats.cpp:199
    #10 0x102aad508 in AnalyzerThread::doRun() analyzerthread.cpp:178
    #11 0x103c36421 in WorkerThread::run() workerthread.cpp:63
    #12 0x10543faad  (QtCore:x86_64+0x2daad)
    #13 0x7fff6d4d8660 in _pthread_body (libsystem_pthread.dylib:x86_64+0x3660)
    #14 0x7fff6d4d850c in _pthread_start (libsystem_pthread.dylib:x86_64+0x350c)
    #15 0x7fff6d4d7bf8 in thread_start (libsystem_pthread.dylib:x86_64+0x2bf8)

Thread T63 created by T0 here:
    #0 0x106e26676 in wrap_pthread_create (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x50676)
    #1 0x1054402c3 in QThread::start(QThread::Priority) (QtCore:x86_64+0x2e2c3)
    #2 0x102ac650d in TrackAnalysisScheduler::TrackAnalysisScheduler(Library*, int, QSharedPointer<ConfigObject<ConfigValue> > const&, AnalyzerModeFlags) trackanalysisscheduler.cpp:83
    #3 0x102ac54df in TrackAnalysisScheduler::createInstance(Library*, int, QSharedPointer<ConfigObject<ConfigValue> > const&, AnalyzerModeFlags) trackanalysisscheduler.cpp:58
    #4 0x103144f7f in AnalysisFeature::analyzeTracks(QList<TrackId>) analysisfeature.cpp:133
    #5 0x1034ce4ce in AnalysisFeature::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) moc_analysisfeature.cc:109
    #6 0x10561fd4a in QMetaObject::activate(QObject*, int, int, void**) (QtCore:x86_64+0x20dd4a)
    #7 0x1034d93e0 in DlgAnalysis::analyzeTracks(QList<TrackId>) moc_dlganalysis.cc:275
    #8 0x1033ed769 in DlgAnalysis::analyze() dlganalysis.cpp:141
    #9 0x1034d8003 in DlgAnalysis::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) moc_dlganalysis.cc:143
    #10 0x10561fd4a in QMetaObject::activate(QObject*, int, int, void**) (QtCore:x86_64+0x20dd4a)
    #11 0x1064bd49e  (QtWidgets:x86_64+0xfa49e)
    #12 0x1064bd336  (QtWidgets:x86_64+0xfa336)
    #13 0x1064be59e in QAbstractButton::mouseReleaseEvent(QMouseEvent*) (QtWidgets:x86_64+0xfb59e)
    #14 0x10640e68e in QWidget::event(QEvent*) (QtWidgets:x86_64+0x4b68e)
    #15 0x1063d36ac in QApplicationPrivate::notify_helper(QObject*, QEvent*) (QtWidgets:x86_64+0x106ac)
    #16 0x1063d64ec in QApplication::notify(QObject*, QEvent*) (QtWidgets:x86_64+0x134ec)
    #17 0x1055ef1b7 in QCoreApplication::notifyInternal2(QObject*, QEvent*) (QtCore:x86_64+0x1dd1b7)
    #18 0x1063d400f in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool) (QtWidgets:x86_64+0x1100f)
    #19 0x10642e5a6  (QtWidgets:x86_64+0x6b5a6)
    #20 0x10642d2a4  (QtWidgets:x86_64+0x6a2a4)
    #21 0x1063d36ac in QApplicationPrivate::notify_helper(QObject*, QEvent*) (QtWidgets:x86_64+0x106ac)
    #22 0x1063d4afb in QApplication::notify(QObject*, QEvent*) (QtWidgets:x86_64+0x11afb)
    #23 0x1055ef1b7 in QCoreApplication::notifyInternal2(QObject*, QEvent*) (QtCore:x86_64+0x1dd1b7)
    #24 0x1059e4cc9 in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) (QtGui:x86_64+0x2ecc9)
    #25 0x1059ccaca in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) (QtGui:x86_64+0x16aca)
    #26 0x10d0cf75d  (libqcocoa.dylib:x86_64+0x2b75d)
    #27 0x10d0d0020  (libqcocoa.dylib:x86_64+0x2c020)
    #28 0x7fff452bb8f0 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ (CoreFoundation:x86_64h+0xa38f0)
    #29 0x7fff4537530b in __CFRunLoopDoSource0 (CoreFoundation:x86_64h+0x15d30b)
    #30 0x7fff4529e34f in __CFRunLoopDoSources0 (CoreFoundation:x86_64h+0x8634f)
    #31 0x7fff4529d7cc in __CFRunLoopRun (CoreFoundation:x86_64h+0x857cc)
    #32 0x7fff4529d032 in CFRunLoopRunSpecific (CoreFoundation:x86_64h+0x85032)
    #33 0x7fff44587d95 in RunCurrentEventLoopInMode (HIToolbox:x86_64+0x2fd95)
    #34 0x7fff44587a0e in ReceiveNextEventCommon (HIToolbox:x86_64+0x2fa0e)
    #35 0x7fff44587883 in _BlockUntilNextEventMatchingListInModeWithFilter (HIToolbox:x86_64+0x2f883)
    #36 0x7fff42837a72 in _DPSNextEvent (AppKit:x86_64+0x41a72)
    #37 0x7fff42fcde33 in -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] (AppKit:x86_64+0x7d7e33)
    #38 0x7fff4282c884 in -[NSApplication run] (AppKit:x86_64+0x36884)
    #39 0x10d0cee0c  (libqcocoa.dylib:x86_64+0x2ae0c)
    #40 0x1055ead3d in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) (QtCore:x86_64+0x1d8d3d)
    #41 0x1055ef8b0 in QCoreApplication::exec() (QtCore:x86_64+0x1dd8b0)
    #42 0x102a4522c in main main.cpp:53
    #43 0x7fff6d1c0014 in start (libdyld.dylib:x86_64+0x1014)

SUMMARY: AddressSanitizer: heap-buffer-overflow PeakFinder.cpp:147 in soundtouch::PeakFinder::getPeakCenter(float const*, int) const
Shadow bytes around the buggy address:
  0x1c4400001760: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c4400001770: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c4400001780: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c4400001790: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c44000017a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x1c44000017b0: 00 00 00 00 00 00 00 00 00 00 00 00[fa]fa fa fa
  0x1c44000017c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c44000017d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c44000017e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c44000017f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c4400001800: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
```

Be-ing commented:
Remove VAMP dependency from Ubuntu AppVeyor dependency list.

rryan committed:
Remove VAMP from SConstruct.

rryan committed:
Remove VAMP from appveyor.yml.

rryan commented:
> Remove VAMP dependency from Ubuntu AppVeyor dependency list.

Oops, bad merge at some point.

rryan commented:
> Build warning:
> 
> ```
> scons: warning: Ignoring missing SConscript 'lin64_build/vamp-plugins/SConscript'
> File "/home/be/sw/mixxx/SConstruct", line 100, in <module>
> ```
> 
> I tried `rm -rf lin64_build .sconf_temp .sconsign*` and the error still appears.

Hm, also a bad merge. Fixed.

rryan committed:
Remove logging and unused constant.

rryan committed:
Remove unused fftw dependency.

Be-ing commented:
Batch analysis goes fwoosh!!
![screenshot from 2018-12-30 10-56-43](https://user-images.githubusercontent.com/9455094/50549385-a1d42600-0c21-11e9-81cf-8d8b319071db.png)

Be-ing commented:
Analyzing 4026 tracks with a fresh settings directory on my quad core Intel Core i7 8550U went from ~7 hours with #1624 to ~1.5 hours with this branch! :raised_hands: LGTM, although I think it would be good if @uklotzde checked that the merge conflicts were resolved correctly before merging this to master.

rryan commented:
> although I think it would be good if @uklotzde checked that the merge conflicts were resolved correctly before merging this to master.

The conflicts were actually pretty minimal, we touched mostly disjoint sets of files -- the only place we conflicted was where we both fixed the same issue in src/library/analysisfeature.h and I just reverted my changes in favor of @uklotzde's.

Be-ing commented:
Okay, I'll merge this now.

Be-ing commented:
Ubuntu Jenkins build failed. I think it's related to this, but I'm not certain:
```
make[1]: Leaving directory '/build/mixxx-2.3.0~alpha~pre'
   dh_install
dh_install: Cannot find (any matches for) "usr/lib/mixxx" (tried in ., debian/tmp)

dh_install: mixxx missing files: usr/lib/mixxx
dh_install: missing files, aborting
debian/rules:34: recipe for target 'binary' failed
make: *** [binary] Error 25
dpkg-buildpackage: error: fakeroot debian/rules binary subprocess returned exit status 2
I: copying local configuration
E: Failed autobuilding of package
I: unmounting dev/ptmx filesystem
I: unmounting dev/pts filesystem
I: unmounting dev/shm filesystem
I: unmounting proc filesystem
I: unmounting sys filesystem
I: cleaning the build env 
I: removing directory /var/cache/pbuilder/build//28332 and its subdirectories
* Build failed.

scons: *** [lin64_build/blah] Exception : Ubuntu package build failed.
Traceback (most recent call last):
  File "/usr/lib/scons/SCons/Action.py", line 1197, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "/home/mixxx/workspace/master-release/architecture/amd64/platform/ubuntu/SConscript", line 1129, in BuildUbuntuPackage
    raise Exception('Ubuntu package build failed.')
Exception: Ubuntu package build failed.
scons: building terminated because of errors.
OSError: [Errno 2] No such file or directory: '.sconsign.dblite':
  File "/usr/lib/scons/SCons/Script/Main.py", line 1376:
    _exec_main(parser, values)
  File "/usr/lib/scons/SCons/Script/Main.py", line 1339:
    _main(parser)
  File "/usr/lib/scons/SCons/Script/Main.py", line 1103:
    nodes = _build_targets(fs, options, targets, target_top)
  File "/usr/lib/scons/SCons/Script/Main.py", line 1313:
    jobs.run(postfunc = jobs_postfunc)
  File "/usr/lib/scons/SCons/Job.py", line 113:
    postfunc()
  File "/usr/lib/scons/SCons/Script/Main.py", line 1310:
    SCons.SConsign.write()
  File "/usr/lib/scons/SCons/SConsign.py", line 116:
    syncmethod()
  File "/usr/lib/scons/SCons/dblite.py", line 157:
    self._os_unlink(self._file_name)
Exception OSError: (2, 'No such file or directory', '.sconsign.dblite') in <bound method dblite.__del__ of <SCons.dblite.dblite object at 0x7fab0419ce10>> ignored
Build step 'Execute shell' marked build as failure
[WARNINGS] Skipping publisher since build result is FAILURE
Recording test results
ERROR: Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?
SSH: Current build result is [FAILURE], not going to run.
Finished: FAILURE
```
https://builds.renegadetech.mixxx.org/job/master-release/architecture=amd64,platform=ubuntu/732/console

uklotzde commented:
> getRemainingFeatures()

I remember that we experienced crashes once in a while in getRemainingFeatures() when running the VAMP plugins in a multi-threaded scenario without additional synchronization!

uklotzde commented:
I also noticed that the legacy SoundTouch beat detector was selected as the default instead of VAMP QM after starting Mixxx with this PR merged.

Be-ing commented:
> I also noticed that the legacy SoundTouch beat detector was selected as the default instead of VAMP QM after starting Mixxx with this PR merged.

That was due to an incorrect merge conflict resolution and fixed in 260071e1513733b18fddc7d8c0efea25a27a8ac8. I suggest doing a clean checkout of current master for testing.
